pkg/runtime/unstructured/converter.go

```diff
@@ -0,0 +1,234 @@
+			// encode it first.
+			data, err := json.Marshal(sv.Interface())
+			if err != nil {
+				return fmt.Errorf("Error encoding to json")
+			}
+			ret := unmarshal.Call([]reflect.Value{reflect.ValueOf(data)})[0].Interface()
+			if ret != nil {
+				return ret.(error)
+			}
+			return nil
+		}
+	}
+
+	switch dt.Kind() {
+	case reflect.Map:
+		return c.mapFromUnstructured(sv, dv)
+	case reflect.Slice:
+		return c.sliceFromUnstructured(sv, dv)
+	case reflect.Ptr:
+		return c.pointerFromUnstructured(sv, dv)
+	case reflect.Struct:
+		return c.structFromUnstructured(sv, dv)
+	case reflect.Interface:
+		return c.interfaceFromUnstructured(sv, dv)
+	default:
+		return fmt.Errorf("Unrecognized type: %v", dt.Kind())
+	}
+}
+
+func fieldNameFromField(field *reflect.StructField) string {
+	jsonTag := field.Tag.Get("json")
+	if len(jsonTag) == 0 {
+		// FIXME: This should start with small letter.
+		return field.Name
+	}
+	return strings.Split(jsonTag, ",")[0]
+}
+
+func unwrapInterface(v reflect.Value) reflect.Value {
+	for v.Kind() == reflect.Interface {
+		v = v.Elem()
+	}
+	return v
+}
+
+func (c *Converter) mapFromUnstructured(sv, dv reflect.Value) error {
+	st, dt := sv.Type(), dv.Type()
+	if st.Kind() != reflect.Map {
+		return fmt.Errorf("Cannot restore map from %v", st.Kind())
+	}
+
+	if !st.Key().AssignableTo(dt.Key()) && !st.Key().ConvertibleTo(dt.Key()) {
+		return fmt.Errorf("Cannot copy map with non-assignable keys: %v %v", st.Key(), dt.Key())
+	}
+
+	if sv.IsNil() {
+		dv.Set(reflect.Zero(dt))
+		return nil
+	}
+	dv.Set(reflect.MakeMap(dt))
+	for _, key := range sv.MapKeys() {
+		value := reflect.New(dt.Elem()).Elem()
+		if val := unwrapInterface(sv.MapIndex(key)); val.IsValid() {
+			if err := c.fromUnstructured(val, value); err != nil {
+				return err
+			}
+		} else {
+			value.Set(reflect.Zero(dt.Elem()))
+		}
+		if st.Key().AssignableTo(dt.Key()) {
+			dv.SetMapIndex(key, value)
+		} else {
+			dv.SetMapIndex(key.Convert(dt.Key()), value)
+		}
+	}
+	return nil
+}
+
+func (c *Converter) sliceFromUnstructured(sv, dv reflect.Value) error {
+	st, dt := sv.Type(), dv.Type()
+	if st.Kind() == reflect.String && dt.Elem().Kind() == reflect.Uint8 {
+		// We store original []byte representation as string.
+		// This conversion is allowed, but we need to be careful about
+		// marshaling data appropriately.
+		if len(sv.Interface().(string)) > 0 {
+			marshalled, err := json.Marshal(sv.Interface())
+			if err != nil {
+				return fmt.Errorf("Error encoding to json: %v", err)
+			}
+			var data []byte
+			err = json.Unmarshal(marshalled, &data)
+			if err != nil {
+				return fmt.Errorf("Error decoding from json: %v", err)
+			}
+			dv.SetBytes(data)
+		} else {
+			dv.Set(reflect.Zero(dt))
+		}
+		return nil
+	}
+	if st.Kind() != reflect.Slice {
+		return fmt.Errorf("Cannot restore slice from %v", st.Kind())
+	}
+
+	if sv.IsNil() {
+		dv.Set(reflect.Zero(dt))
+		return nil
+	}
+	dv.Set(reflect.MakeSlice(dt, sv.Len(), sv.Cap()))
```

Comment:
Copying from disappeared comment:
>This is the bit that I was referring to possibly being able to avoid - we can use the target slice as the destination and not have to invoke MakeSlice.

I'm afraid I still don't understand. Can you please clarify a bit more?